package cetty.protobuf.service;

option java_package = "cetty.protobuf.service";
option java_outer_classname = "ServiceProto";

import "google/protobuf/descriptor.proto";

enum MessageType {
  REQUEST  = 1;
  RESPONSE = 2;
  ERROR    = 3;
}

enum ErrorCode {
  WRONG_PROTO      = 1;
  NO_SERVICE       = 2;
  NO_METHOD        = 3;
  INVALID_REQUEST  = 4;
  INVALID_RESPONSE = 5;
}

message ServiceMessage {
  required MessageType type = 1;
  required fixed64       id = 2;

  optional string   service = 3;
  optional string    method = 4;
  
  optional ErrorCode  error = 5;
  
  optional bytes    request = 6;
  optional bytes   response = 7;
}

message Empty {
}

extend google.protobuf.MethodOptions {
  optional bool idempotent = 1111;
  optional bool  no_return = 1112;
}
